rust/connlib/snownet/tests/lib.rs

```diff
@@ -189,3 +216,210 @@ fn s(socket: &str) -> SocketAddr {
+        local: SocketAddr,
+        from: SocketAddr,
+        packet: &[u8],
+        now: Instant,
+        buffer: &'s mut [u8],
+    ) -> Result<Option<(u64, MutableIpPacket<'s>)>, snownet::Error> {
+        match self {
+            EitherNode::Client(n) => n.decapsulate(local, from, packet, now, buffer),
+            EitherNode::Server(n) => n.decapsulate(local, from, packet, now, buffer),
+        }
+    }
+
+    fn handle_timeout(&mut self, now: Instant) {
+        match self {
+            EitherNode::Client(n) => n.handle_timeout(now),
+            EitherNode::Server(n) => n.handle_timeout(now),
+        }
+    }
+}
+
+impl TestNode {
+    pub fn new(span: Span, node: impl Into<EitherNode>) -> Self {
+        let now = Instant::now();
+        TestNode {
+            node: node.into(),
+            span,
+            progress_count: 0,
+            time: now,
+            received_packets: vec![],
+            buffer: Box::new([0u8; 10_000]),
+            connection_state: HashMap::default(),
+        }
+    }
+
+    fn is_connected_to(&self, id: u64) -> bool {
+        self.connection_state.get(&id).copied().unwrap_or_default()
+    }
+
+    fn add_local_host_candidate(&mut self, socket: &str) {
+        self.span
+            .in_scope(|| self.node.add_local_host_candidate(socket))
+    }
+}
+
+fn handshake(client: &mut TestNode, server: &mut TestNode) {
+    let client_node = &mut client.node.as_client_mut().unwrap();
+    let server_node = &mut server.node.as_server_mut().unwrap();
+
+    let offer = client.span.in_scope(|| {
+        client_node.new_connection(
+            1,
+            HashSet::default(),
+            HashSet::default(),
+            client.time,
+            client.time,
+        )
+    });
+    let answer = server.span.in_scope(|| {
+        server_node.accept_connection(
+            1,
+            offer,
+            client_node.public_key(),
+            HashSet::default(),
+            HashSet::default(),
+            server.time,
+        )
+    });
+    client
+        .span
+        .in_scope(|| client_node.accept_answer(1, server_node.public_key(), answer, client.time));
+}
+
+fn progress(a1: &mut TestNode, a2: &mut TestNode) {
+    let (f, t) = if a1.progress_count % 2 == a2.progress_count % 2 {
+        (a2, a1)
+    } else {
+        (a1, a2)
+    };
+
+    t.progress_count += 1;
+    if t.progress_count > 100 {
+        panic!("Test looped more than 100 times");
+    }
+
+    while let Some(v) = t.span.in_scope(|| t.node.poll_event()) {
+        match v {
+            Event::SignalIceCandidate {
+                connection,
+                candidate,
+            } => f.node.add_remote_candidate(connection, candidate, f.time),
+            Event::ConnectionEstablished(id) => {
+                *t.connection_state.entry(id).or_default() = true;
+            }
+            Event::ConnectionFailed(id) => {
+                *t.connection_state.entry(id).or_default() = false;
+            }
```

Comment:
I think I want to refactor this so `snownet` exposes whether or not you are connected.